### DISCUSSION TITLE
Flip CC polarity

### DISCUSSION THREAD

**Discussion by Saga-Daroxirel** [FluidSynth/fluidsynth#1046] [Q&A] [answered]
I was wondering if there's any way to flip the polarity of an input. I have a pedal that looks like it's registered to CC 64, but pressing the pedal is 0 and releasing it is 127, making the sound sustain when I'm _not_ holding it. It looks like I might be able to reverse it using router rules, but I haven't been able to find enough information to figure them out.  Is there any way to change it without downloading and manually compiling the core DLL with the input inverted?

**Comment by mawe42** [comment, marked as answer]
Hi,

the router isn't very well documented, I'm afraid. Assuming you can use the FluidSynth shell (either via telnet, command-line or config file), you can set up the router as follows.

The most important thing to understand is that clearing the router (so you can reconfigure it) means that **no messages** get passed anymore. To restore the default router where all messages are sent unchanged, you need to create a default route for each of the 6 event types:
```
router_clear

router_begin note
router_end

router_begin cc
router_end

router_begin prog
router_end

router_begin pbend
router_end

router_begin cpress
router_end

router_begin kpress
router_end
```

So after executing these commands, your FluidSynth should act unchanged, all messages pass as they were sent, no messages get dropped. You could also use the `router_defaultl` command instead of the individual default routes per event type as above. But doing it explicity makes it clearer, I think.

If you wanted to complete ignore program change events, you would simply leave out the `router_begin prog, router_end` bit.

Now to achieve your goal of reversing a cc, you should be able to use the following rule:

```
# new CC rule
router_begin cc
# match CCs on all channels, multiply channel num by 1, add 0
router_chan 0 15 1 0
# match only on CC 64 (64 - 64), multiply num by 1, add 0
router_par1 64 64 1 0
# match all values (0 - 127), multiply the value by -1 and add 127.
router_par2 0 127 -1 127
router_end
```

Add this rule above the empty default rule `router_begin cc; router_end` as shown above, then your CC should be reversed. Rules are evaluated from top to bottom, first rule wins.

So putting it all together, the following should work:
```
router_clear

router_begin note
router_end

router_begin cc
router_chan 0 15 1 0
router_par1 64 64 1 0
router_par2 0 127 -1 127
router_end

router_begin cc
router_end

router_begin prog
router_end

router_begin pbend
router_end

router_begin cpress
router_end

router_begin kpress
router_end
```

You can add all this into a text file and load the file using the `-f` command-line switch. Or type them in interactively via the shell.

**Reply by Saga-Daroxirel** [reply]
Thank you so much!
